src/Compilers/Extension/CompilerPackage.cs

```diff
@@ -90,8 +79,41 @@ private void WriteMSBuildFiles(string packagePath, string hiveName)
   </PropertyGroup>
 </Project>";
 
-                WriteMSBuildFile(targetsContent, $@"Microsoft.CSharp.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets");
-                WriteMSBuildFile(targetsContent, $@"Microsoft.VisualBasic.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets");
+            filesToWrite.Add(GetMSBuildRelativePath($@"Microsoft.CSharp.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets"), targetsContent);
+            filesToWrite.Add(GetMSBuildRelativePath($@"Microsoft.VisualBasic.targets\ImportBefore\Roslyn.Compilers.Extension.{hiveName}.targets"), targetsContent);
+
+            // First we want to ensure any Roslyn files with our hive name that we aren't writing -- this is probably
+            // leftovers from older extensions
+            var msbuildDirectory = new DirectoryInfo(GetMSBuildPath());
+            if (msbuildDirectory.Exists)
+            {
+                foreach (var file in msbuildDirectory.EnumerateFiles($"*Roslyn*{hiveName}*", SearchOption.AllDirectories))
+                {
+                    if (!filesToWrite.ContainsKey(file.FullName))
+                    {
+                        file.Delete();
+                    }
+                }
+            }
+
+            try
+            {
+                foreach (var fileAndContents in filesToWrite)
+                {
+                    var parentDirectory = new DirectoryInfo(Path.GetDirectoryName(fileAndContents.Key));
+                    if (!parentDirectory.Exists)
```

Comment:
The `if` is unnecessary here. The Create method is a no-op if the directory exists.